File: packages/ai-core/src/tools/whatsapp.py
import httpx
import os


class WhatsAppTool:
    """Tool for sending WhatsApp messages"""

    def __init__(self):
        self.phone_id = os.getenv("WA_PHONE_ID", "")
        self.access_token = os.getenv("WA_ACCESS_TOKEN", "")
        self.api_url = f"https://graph.facebook.com/v18.0/{self.phone_id}"

    def send_message(self, to: str, text: str) -> bool:
        """Send a WhatsApp message"""
        if not self.phone_id or not self.access_token:
            return False

        try:
            with httpx.Client() as client:
                res = client.post(
                    f"{self.api_url}/messages",
                    json={
                        "messaging_product": "whatsapp",
                        "to": to,
                        "type": "text",
                        "text": {"body": text}
                    },
                    headers={
                        "Authorization": f"Bearer {self.access_token}",
                        "Content-Type": "application/json"
                    },
                    timeout=10.0
                )
                return res.status_code == 200
        except Exception:
            return False

    def send_buttons(self, to: str, text: str, buttons: list) -> bool:
        """Send WhatsApp message with interactive buttons"""
        if not self.phone_id or not self.access_token:
            return False

        try:
            with httpx.Client() as client:
                res = client.post(
                    f"{self.api_url}/messages",
                    json={
                        "messaging_product": "whatsapp",
                        "to": to,
                        "type": "interactive",
                        "interactive": {
                            "type": "button",
                            "body": {"text": text},
                            "action": {
                                "buttons": [
                                    {"type": "reply", "reply": {"id": btn["id"], "title": btn["title"]}}
                                    for btn in buttons[:3]
                                ]
                            }
                        }
                    },
                    headers={
                        "Authorization": f"Bearer {self.access_token}",
                        "Content-Type": "application/json"
                    },
                    timeout=10.0
                )
                return res.status_code == 200
        except Exception:
            return False

File: apps/chat/src/whatsapp.ts
import axios from 'axios';

const WA_PHONE_ID = process.env.WA_PHONE_ID || '';
const WA_ACCESS_TOKEN = process.env.WA_ACCESS_TOKEN || '';
const WA_API_URL = `https://graph.facebook.com/v18.0/${WA_PHONE_ID}`;

export interface MessageButton {
  id: string;
  title: string;
}

export interface MessageTemplate {
  name: string;
  language: string;
  components?: any[];
}

export class WhatsAppClient {
  private async send(payload: any) {
    if (!WA_PHONE_ID || !WA_ACCESS_TOKEN) {
      console.log('WhatsApp not configured, skipping send');
      return;
    }

    await axios.post(
      `${WA_API_URL}/messages`,
      payload,
      {
        headers: {
          Authorization: `Bearer ${WA_ACCESS_TOKEN}`,
          'Content-Type': 'application/json'
        }
      }
    );
  }

  async sendText(to: string, text: string) {
    await this.send({
      messaging_product: 'whatsapp',
      to,
      type: 'text',
      text: { body: text }
    });
  }

  async sendButtons(to: string, text: string, buttons: MessageButton[]) {
    if (buttons.length === 0) {
      await this.sendText(to, text);
      return;
    }

    await this.send({
      messaging_product: 'whatsapp',
      to,
      type: 'interactive',
      interactive: {
        type: 'button',
        body: { text },
        action: {
          buttons: buttons.slice(0, 3).map(btn => ({
            type: 'reply',
            reply: { id: btn.id, title: btn.title }
          }))
        }
      }
    });
  }

  async sendImage(to: string, imageUrl: string, caption?: string) {
    await this.send({
      messaging_product: 'whatsapp',
      to,
      type: 'image',
      image: { link: imageUrl, caption: caption || '' }
    });
  }

  async sendTemplate(to: string, template: MessageTemplate) {
    await this.send({
      messaging_product: 'whatsapp',
      to,
      type: 'template',
      template
    });
  }

  async markAsRead(messageId: string) {
    await this.send({
      messaging_product: 'whatsapp',
      status: 'read',
      message_id: messageId
    });
  }
}
